File: LC/java/Word Ladder.java
public class Solution {
    public int ladderLength(String start, String end, Set<String> dict) {
        if(dict == null || dict.size()==0){
            return 0;
        }
        Queue<String> q = new LinkedList<String>();
        q.add(start);
        dict.remove(start);
        
        int res=1;
        
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i<size;i++){
                String tmp = q.poll();
                for(char c ='a'; c<='z';c++){
                    for(int j = 0; j<tmp.length();j++){
                        if(c==tmp.charAt(j)){
                            continue;
                        }
                        String t = replace(tmp,j,c);
                        if(t.equals(end)){
                            return res+1;
                        }
                        if(dict.contains(t)){
                            dict.remove(t);
                            q.add(t);
                        }
                    }
                }
            }
            res++;
        }
        return 0;
    }
    
    public String replace(String s, int i, char a){
        char[] c = s.toCharArray();
        c[i] = a;
        return new String(c);
    }
}

File: LC/python/Word Ladder.py
class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return an integer
    def ladderLength(self, start, end, dict):
        q1 = [start]
        n = 1
        dict.add(end)
        
        while q1:
            q2 = []
            while q1:
                tmp = q1.pop(0)
                for i in range(len(tmp)):
                    for j in 'abcdefghijklmnopqrstuvwxyz':
                        word = tmp[:i]+j+tmp[i+1:]
                        if word in dict:
                            q2.append(word)
                            dict.remove(word)
            if end in q2:
                return n+1
            else:
                n += 1
            q1 = q2
        return 0
         
